### ISSUE TITLE
Propagating context information to child-tasks and remote calls?

### ISSUE THREAD

**Issue by oschulz** [JuliaLang/julia#35757]
@tkf and me have been discussing ways to propagate information about available workers (or resources in general) in distributed hierarchical computations:

https://discourse.julialang.org/t/propagation-of-available-assigned-worker-ids-in-hierarchical-computations

Adding resource arguments to every function call would be impractical, and using Cassette & friends to add add context by rewriting the whole computation would be very heavy-handed, since it might tough large code stacks (and may also be not be a complete solution when remote calls are involved?).

Could we add something like a `context` field to `Task`, in addition to the `storage` field - with the difference that `context` is automatically propagated to spawned tasks and remote calls? Adding the possibility to propagate a context through a computation in this fashion could also be useful for other use cases too, I think.

**Comment by vtjnash** [comment]
C.f. #34543

**Comment by tkf** [comment]
Thanks for opening the issue.

I'm actually against copying `task_local_storage` API.  I think something similar to [`contextvars` in Python](https://docs.python.org/3/library/contextvars.html) (see [PEP 567](https://www.python.org/dev/peps/pep-0567/) for discussion) would be better:

```julia
const var1 = ContextVar(:var1, 42)  # with default
# possible sugar: @contextvar var1 = 42

const var2 = ContextVar{Lockable{Vector{Int}}}(:var2)  # without default, with type
# possible sugar: @contextvar var2::Lockable{Vector{Int}}
# Lockable from https://github.com/JuliaLang/julia/pull/34400

const var3 = ContextVar(:var3)  # equivalent to ContextVar{Any}(:var3)
# possible sugar: @contextvar var3


function f()
    @sync @async @show var1[] # var1[] => 42
    var1[] = 0
    @sync @async @show var1[] # var1[] => 0

    setting(var1 => 1) do
        @show var1[] # var1[] => 1
    end
    @show var1[] # var1[] => 0
end
```

In contrast to `task_local_storage(:var)`:

1. `var[]` can be inferred
2. `var` is forced to be namespaced (i.e., it has to exist in some module name space)
3. `var` can be backed up by an efficient concrete key type (e.g., UUID)
4. `var` allows small-size optimization when the value type can be inlined into the context storage

We may want to use something like HAMT (as in PEP 567) for the context storage.

---

A possible implementation of `ContextVar` may be something like

```julia
struct ContextVar{T}
    name::Symbol  # only for human readability
    key::UUID     # (for example)
    default::T    # or `Union{Some{T},Nothing}`?
    has_default::Bool
end

getindex(var::ContextVar{T}) where {T} =
    if var.has_default
        get(task_local_storage(:CONTEXT), var.key, var.default)
    else
        task_local_storage(:CONTEXT)[var.key]
    end :: T
```

**Comment by tkf** [comment]
Here is a proof-of-concept and its usage: https://github.com/tkf/ContextVariables.jl/blob/ebc2b550e2177c4ebc059b6974f1ea36309d7f4f/test/runtests.jl#L14-L29

**Comment by oschulz** [comment]
Oh, nice! Will this work across remote-call boundaries?

Maybe use `with_context` instead of `with_variables`? Variable is such a generic term. :-)

**Comment by tkf** [comment]
Yeah, `with_context` sounds like a much better name. :)

> Will this work across remote-call boundaries?

I think it'd work with local context variables (i.e., captured in closures) but not with global const context variables.  But I think it's possible in principle.  We need to tweak Distributed (or you need to propagate it manually) though.

Notes: Global const doesn't work because I'm using `uuid4` (which is based on the global RNG) to generate the key.  I can use `uuid5` to make the key deterministic (i.e., hashing the package's UUID, module namespaces, and the variable name).  We need a macro to make this easier.

**Comment by c42f** [comment]
Yes we need something better here, thanks again for opening this issue! I feel an alternative name for this issue could be "taking dynamic scope seriously" ;-) Though perhaps "context variable" is a more straightforward name for this.

If we somehow had efficient codegen for context variables, there's a lot of interesting language facilities which can be improved based on this. For example
* It could replace the special cases for the logger state in the existing logging system
* Error handling with an "algebraic effects" or "condition system" - see some discussion at #7026
* Deterministic task local random number generators https://github.com/JuliaLang/julia/pull/34852 (see also https://discourse.julialang.org/t/parallel-mersenne-twister/27567). Interesting and very difficult wrinkle: serving the RNG use case requires running code at task creation time, not just efficient access.

At https://github.com/JuliaLang/julia/issues/35690#issuecomment-623904330 I speculated about one possible way to specialize code based on context variables. As noted in the OP, specializing a whole call stack based on the presence/type of a context variable is just really heavy handed. But on the other hand, I'd speculate that specializing a leaf function (or a few innermost inlined frames) on a context var could be very powerful, and possibly of acceptable cost.

Getting this idea to be sound and practical within inference seems quite tricky. I guess it would require a context calling convention to allow inference to reason about the innermost frames in a systematic way. If it actually panned out, I'm imagining the compiler could hoist context variable access out of innermost loops and across inlined function calls.

Kotlin seems to have some interesting APIs around context and coroutines (and they have taken structured concurrency seriously!) so we might learn something from examining that closer https://kotlinlang.org/docs/reference/coroutines/coroutine-context-and-dispatchers.html

**Comment by oschulz** [comment]
> Getting this idea to be sound and practical within inference seems quite tricky.

I guess it would depend on how deep we want to take it. Context attached to tasks could be done with just a few small, lightweight changes. I think that would cover many use cases already.

Context at a deeper level, propagating through all function calls will come with a higher cost, I guess - but there are quite a few uses cases that would need that, of course, though I think less concerned with scheduling and resource allocation questions.

**Comment by tkf** [comment]
@c42f What do you think about the design of `ContextVar{T}` I posted above?  I think it solves the problem of inferrability.  There is still a dynamic code _inside_ `getindex(::ContextVar{T})` but LICM'ing it (or maybe rather `get(::ContextVar{T}) :: Union{Some{T}, Nothing}` which won't throw) sounds much easier with the current infrastructure than implementing the new effect specialization facility.  I think `ContextVar{T}` is a neat design that completely eliminates the type-instability on the user side of the code.

Also, I don't think it's possible to eliminate _some_ dynamic typing.  Since we need to allow arbitrary value types, and we can't specialize for the _whole_ set of the context variables currently set (i.e., equivalent to using `NamedTuple` as the context), the `valtype` would be `Any`.  So I think type-stabilizing at use-site is a good strategy.

(I wonder if it can be generalized to the strategy for implementing dynamically scoped effect handler. That is to say, type information is statically "hoisted out" to the global/lexical scope and then there is a tiny bits of run-time information invoking those pre-defined entrypoints.)

I also thought about supporting the deterministic parallel RNG.  It probably is possible via implementing some kind of hook system via `AbstractContextVar`.  But I fear there will be too much dynamic dispatches on the task creation.  Though it looks like [Kotlin's context system allows user-defined `restoreThreadContext`/`updateThreadContext`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-thread-context-element/index.html)?  I wonder how they make it efficient.  Is it doable because Kotlin is statically typed or something?

**Comment by c42f** [comment]
> What do you think about the design of `ContextVar{T}` I posted above? I

Well, I like it! It's simple and can be implemented right away with only some small changes to the runtime. My comments above about effect systems, etc, are all pretty much pie-in-the-sky speculation at this stage, though it would be nice to have a rough feeling for where things might go.

So anyway, if we provided `ContextVar`, that at least gives us strongly typed task local storage which would be reasonably efficient, convenient and solve some problems with context propagation which are currently unsolvable for users. It also doesn't seem too far-fetched to teach the Julia optimizer about the scoping semantics so that some form of LICM will work in the medium term. Specialization based on type of context vars could eventually be introduced as an optimization (if it's even feasible).

**Comment by oschulz** [comment]
> I also thought about supporting the deterministic parallel RNG

That's actually something I already have in BAT.jl

https://github.com/bat/BAT.jl/blob/master/src/rngs/rng_init.jl

This provides reproducible for hierarchical parallel applications, via hierarchical partitioning of counter-based RNGs .

I'll make that more widely available via ParallelProcessingTools.jl. It's just in BAT.jl currently because I needed it fast and didn't have time to do a nice API at the time. I'll get on it.

**Comment by oschulz** [comment]
Context variables would be super-useful to propagate those RNG partitions!

**Comment by JeffBezanson** [comment]
I think the core problem is the cost of the key lookup. I believe it would be much too slow for something like RNG.

**Comment by oschulz** [comment]
> I think the core problem is the cost of the key lookup. I believe it would be much too slow for something like RNG.

Well, I guess in many use cases, the task would get it's RNG once, and then pass it on to the functions it calls as an explicit parameter. And then pass partitions of that RNG to sub-tasks, when they spawn. I really wasn't trying to go for a full resource-injection solution with this issue, just something on the task-level for low/medium-frequency access.

The RNG-story is a bit different from the worker-availability story, though. Which RNG is used for which part of the computation shouldn't depend on number or workers/threads/tasks, for reproducibility, so that often may need to be handled explicitly.

**Comment by tkf** [comment]
> My comments above about effect systems, etc, are all pretty much pie-in-the-sky speculation at this stage, though it would be nice to have a rough feeling for where things might go.

Yeah, I do like pie-in-the-sky speculations and am very interested in effect systems!  Also, I think context variable API helps implementing POC effect handlers, even though it won't be as efficient as we'd want.  Some effect handlers don't require too much optimizations when the overhead of the handler itself is large (e.g., a process+thread pool abstracting Distributed and Threads).  I think it'd be nice to have a building block for playing with effect handler interfaces and assessing programming experience with it, before start thinking about optimizing the hell out of it (though it's also important to think about compiler-friendly interface at the same time).

**Comment by tkf** [comment]
> This provides reproducible random numbers for hierarchical parallel applications, via hierarchical partitioning of counter-based RNGs .

@oschulz It's awesome that you already have counter-based RNGs integrated into a parallel program! But don't you need to do something at `@spawn`?  I guess your other comment confirms that?:

> And then pass partitions of that RNG to sub-tasks, when they spawn.

If so, don't you need to create custom `@spawn`-like API anyway?  I'm trying to understand how context variables can be useful in implementing deterministic parallel RNGs in "user land." I _thought_ it'd be rather useless.

**Comment by c42f** [comment]
> I think the core problem is the cost of the key lookup. I believe it would be much too slow for something like RNG.

Yeah. To make this fast enough we need to be able to hoist the load of the `ContextVar` for the global RNG out of any inner loop, and also have a fixed concrete type for that RNG. (The fixed type is somewhat limiting, but no worse than what we have now.) I feel like it should be feasible to model dynamic scope in the optimizer if all setting and getting of ContextVars goes through a well defined interface? That could include eliding the ContextVar load for inlined frames where the ContextVar store is visible. And even eliding the store if the frame is a leaf?

**Comment by oschulz** [comment]
> @oschulz It's awesome that you already have counter-based RNGs integrated into a parallel program! But don't you need to do something at @spawn? I guess your other comment confirms that?:

Not for the RNG, no, because all functions I use take an RNG as an explicit argument. Since this is pretty much a standard in most of the Julia ecosystem, it's easy to do that consistently, also with the third-party packages I use. In the beginning, I also had explicit function arguments for resources like threads in BAT.jl, but that was unwieldy and I got rid it when partr came around.

Like I wrote, RNG distribution/partitioning can't always be automatic if the computation should be reproducible independing of parallel execution strategy. But it would still be great to have the option of propagating the RNG via context - user's don't always like to have to pass the RNG explicitly - but at points in between, algorithms that distribute computation will need to do some explicit RNG handling/partitioning.

**Comment by tkf** [comment]
> But it would still be great to have the option of propagating the RNG via context

Hmm....  I'm not sure if I understand. Let's consider this snippet:

```julia
@contextvar state = 0

function demo()
    @sync begin
        @async println("at task1: ", state[])
        @async println("at task2: ", state[])
        state[] += 1
        println("at task0: ", state[])
    end
end
```

Then calling `demo()` will print

```
at task0: 1
at task1: 0
at task2: 0
```

in some order.  If `state` were the state of RNG, `task1` and `task2` will have the same stream.  That's not what you want, right?

`task1` and `task2` can call your functions that use RNG correctly.  But, unless every use of `@async`/`@spawn` is aware of the RNG state that is in the task context, you can't be sure that you have "independent" streams.

**Comment by tkf** [comment]
I think there are two design questions:

(1) Should Distributed handle it automatically? That is to say, should context variable values be restricted to serializable values? It may be handy to put non-serializable states like files and locks in it so I don't think we should add this restriction. But this would make it impossible to implement automatic handling by Distributed. (Note that users still can implement context propagation by wrapping Distributed API and manually propagating known-to-be-safe context variables.)

A semi-automatic way may be to add an overloadable function API `Distributed.is_serializable_context_value(x) :: Bool` or something.  This way, Distributed can copy the context for objects that are marked serializable-safe explicitly.

(2) Should it be possible to list context variables currently set? With the current design (only storing a mapping `uuid => value` in `Task`), it's not possible to get this. We can store a mapping `uuid => (var, value)` in `Task` or something equivalent to get the list but I'm not sure about the performance impact of this idea.

My preference is to add the context variable API without these features and see if we can get away without implementing them.

**Comment by tkf** [comment]
OK so here is a full set of API I'd propose https://tkf.github.io/ContextVariables.jl/dev/. It includes a quick tutorial.

**Comment by oschulz** [comment]
> If state were the state of RNG, task1 and task2 will have the same stream. That's not what you want, right?

Oh, yes - we will of course, in general, need variants of `@spawn`, `@remotecall`, `@async` and so on that allow for specifying a partial new context (which should be merged with the current context before sending it on).

With my partitioned RNGs, the situations is a bit different, though. An [RNGPartition](https://github.com/bat/BAT.jl/blob/557acc91f5591877b7e60745426c0fbaa24f4beb/src/rngs/rng_init.jl#L20) can be passed on as it is, though, because the receiving end will instantiate the actual RNG (e.g. via `AbstractRNG(rngpart::RNGPartition{R}, i::Integer)`) it should use based on which part of the calculation it will handle next, not based on it's position in the worker/task hierarchy.

However, parts of the calculation that are not aware of RNGs (e.g. some 3rd party package), should pass on the RNG or RNG partition unchanged automatically, if possible.

> Should Distributed handle it automatically?

I think yes.

The idea would be that a hierarchical calculation will, at certain points, need to control how resources (workers, RNG, etc.) are to be partitioned

**Comment by c42f** [comment]
> OK so here is a full set of API I'd propose https://tkf.github.io/ContextVariables.jl/dev/. It includes a quick tutorial.

I haven't had time to read all the implementation yet, but the API looks nice. I particularly like the section on data races. I think you've nailed the correct semantics there: the runtime must ensure that the ContextVar get/set is data race free between Tasks, but the use of any context var values will need to be made threadsafe by the user.

For implementation of storage we could add an extra `context` field on `Task`, which, similar to `logstate`, is copied at task creation but is otherwise free to use in this work. Then store something like `ImmutableDict` in that context field. (Ideally `logstate` would eventually be removed and replaced with a ContextVar.)

> Note that users still can implement context propagation by wrapping Distributed API and manually propagating known-to-be-safe context variables

This seems like it can be worked around by the application author but it's going to be ugly; in principle they can make a big list of `ContextVar`s they care about.  But it seems like libraries making use of `ContextVar` internally won't be able to compose with libraries using `Distributed`?

It's a tricky problem because in general remote calls can't know which of the context vars will actually be used. So should it be an error to have a contextvar installed which can't be serialized, but which won't be used on the remote side? I would think not.  Perhaps it could be made to work by allowing the remote call and sending all context vars, but with the value of any non-serializable context var poisoned so that any use of it on the remote side will result in an exception.

> Should it be possible to list context variables currently set

I agree we might be get away without this. The proposed API seems to be one where module authors are likely to create a `ContextVar` for their own internal purposes within a module. So if nobody else can list those it seems fine.

On the other hand, it would be *really* useful to be able to list the attached `ContextVar`s during debugging. But for that a separate global lookup table based on uuid might suffice anyway?

**Comment by tkf** [comment]
> The idea would be that a hierarchical calculation will, at certain points, need to control how resources (workers, RNG, etc.) are to be partitioned

@oschulz Yeah, I agree it'd be nice to be automatic for these cases. But, as I mentioned, it some undesirable consequences.  For example, some objects like files and locks can't cross process boundaries.  It'd be problematic if a user accidentally put huge arrays in the context variable.

(Another way may be to use `default_worker_pool` to handle context variables that describe computation resources. Discussed below.)

> For implementation of storage we could add an extra `context` field on `Task`, which, similar to `logstate`, is copied at task creation but is otherwise free to use in this work.

@c42f Yeah, I actually have already implemented it like this :) https://github.com/tkf/julia/commits/ctxvars

> It's a tricky problem because in general remote calls can't know which of the context vars will actually be used.

One solution may be to make `Distributed.default_worker_pool` backed up by a context-local variable.  This way, it's (probably) possible to implement context propagation across process boundaries by implementing a custom worker pool on user space.  Then such custom worker pool can decide what context variables to propagate.  I'm not sure if worker and worker pool are designed to allow this, but, if we are to implement context propagation across processes, it sounds like a nice hook point to make it work. (I guess it's kind of like how LoggingExtras.jl builds composable stack on the user land.)

> On the other hand, it would be _really_ useful to be able to list the attached `ContextVar`s during debugging. But for that a separate global lookup table based on uuid might suffice anyway?

Yeah, having an optional lookup table for debugging sounds like a good idea.

**Comment by tkf** [comment]
I opened PR #35833 to add this API so that it'd be easier to comment on implementation/design specific to the API I'm proposing.

**Comment by oschulz** [comment]
> OK so here is a full set of API I'd propose

I'm not so sure that we should directly inject variables into the scope of the child-tasks. There may be name clashes and it's a potential security issue too - mainly I'm worried about name conflicts though. I think it would be better if context variables were retrieved with an explicit mechanism.

**Comment by oschulz** [comment]
> Yeah, I agree it'd be nice to be automatic for these cases. But, as I mentioned, it has some undesirable consequences.

Hm, I would hope those problems can be overcome somehow - if only by the user being reasonable.

However, for the original scope of this issue - propagation of available workers - we'd most certainly want fully automatic propagation, so the the information is not lost when using packages that don't use the context-API to spawn local/remote tasks.

**Comment by tkf** [comment]
Thanks for having a look into the API docs.

> There may be name clashes

Perhaps the documentation should be fixed to emphasize this but there will be no name clashes, by design.  (Unless you can invoke a collision of UUIDs.)

> However, for the original scope of this issue - propagation of available workers

The set of available workers is a very dynamic information and I don't think propagating it via "static" mechanism like context variable is a good idea.  It would mean that a function doesn't get any update after it is called via `@spawn` or `remotecall`.  I think a better approach is to create a worker pool abstraction [*1] and then propagate the currently active worker pool using the context variable.

This is why @c42f and I are discussing the effect system here.  Task scheduler is a special case of the effect handler (https://github.com/JuliaLang/julia/issues/33248#issuecomment-572877801) and we need dynamic scoping to implement this (or rather a small subset of effect handler that is still enough for task scheduler).  Context variable just provides dynamic scoping and some effect handlers can be implemented on top of it.

[*1] I guess it would need to implement something like the work-stealing approach on top of Threads and Distributed.

**Comment by oschulz** [comment]
There's actually precedent for task inheriting information from each other: Tasks do inherit `logstate` from their parent. I think it would make a lot of sense if tasks also had `resources` (should probably be a dict) that they inherit, and that the user can add content too.

**Comment by oschulz** [comment]
See https://github.com/JuliaLang/julia/blob/9a712ae0cedc7b5a8684ddf50988a0f23acecd9f/src/task.c#L579

**Comment by tkf** [comment]
Yes, that's how it's implemented in #35833.

**Comment by oschulz** [comment]
Oh, sure - I was mostly lending moral support - should have posted it on #35833.

**Comment by tkf** [comment]
Ah, sorry, I misread your comment.  Yes, it'd be nice to expand what is done for loggers to something more general.

(Actually, your comment makes me realize that it would have been _much_ easier to write the proof-of-concept using a custom logger than using generated function...)

**Comment by c42f** [comment]
Yeah, I'm responsible for adding that `logstate` thing :-) It's currently an odd special case, but really shouldn't be.

**Comment by oschulz** [comment]
Yes, logging is still a bit inconsistent, in some ways ... :-) But I very much like that by-task logstate, and the fact that it's inherited. Do you think you could propagate that to remote calls as well?

Maybe this whole thing could be generalized, actually - in a way, a logstate is a kind of resource that is propagated, right?

**Comment by oschulz** [comment]
What's important is that there must be a way to override the automatic inheritance during spawns and remote calls, so that a task can choose to propagate different values for some resources in certain cases - a subset of workers that the child task is allowed to use, less detailed logging at lower levels of the computation, etc. I think this kind of general approach could be very powerful, if available consistently.

**Comment by c42f** [comment]
> there must be a way to override the automatic inheritance during spawns and remote calls, so that a task can choose to propagate different values for some resources in certain cases

I don't think this is directly related to `@spawn`, but an orthogonal feature which is part of the dynamically scoped context system (`with_context` in the PR). You may be using a third party algorithm which happens to use `@spawn` as an implementation detail, but that third party package can't know about the larger application's logging requirements.

**Comment by oschulz** [comment]
But wouldn't it be important that for code that is not aware of this, context is still propagated automatically, both locally and remotely?

**Comment by tkf** [comment]
> context is still propagated automatically

That's how it works in #35833, at least locally.  It may be possible to get semi-automated context propagation to the remote workers but we need to get multi-threading part right to get there.  I've already designed the internal so that it is Distributed-friendly, even for multi-node scenario.  So, it is already possible to manually bridge contexts across remote calls with automated propagation across threads.

However, as I've explained in https://github.com/JuliaLang/julia/issues/35757#issuecomment-626967900, "a set of available workers" is a very dynamic concept (each `@spawn`/`remotecall` changes the set).  Also, the context does not propagate backwards from the child tasks to the parent task (this is a common behavior AFAICT for this kind of system).  See the [documentation](https://tkf.github.io/ContextVariables.jl/dev/) and especially the [Concurrent access](https://tkf.github.io/ContextVariables.jl/dev/#Concurrent-access-1) section.  So, it cannot be used as a communication mechanism for dynamic global resource.  I think you need to handle this with a very different mechanism.  Maybe a global state with a lock or some kind of coordinator task listening to channels.  (I'm just mentioning this as, IIUC, that's your primary goal.)

**Comment by tkf** [comment]
> it cannot be used as a communication mechanism for dynamic global resource

I said `with_context` is like dynamic `let` in the documentation. But I just noticed that that's actually not true because of this.

**Comment by oschulz** [comment]
> However, as I've explained in #35757 (comment), "a set of available workers" is a very dynamic concept (each @spawn/remotecall changes the set)

In some situations for sure, where you have some central scheduling. But in other situations, one may want to split computations on a longer term to run complete algorithms (that do their own distribution too) on subsets of the available worker, with each of them handled by a "master" instance or so.

**Comment by oschulz** [comment]
Basically, it should be possible to partition the worker, with partitions that live longer than a single call. Each partition may cache a specific subset of input data, for example - still, partitions may not be fixed for the entire Julia session.

**Comment by tkf** [comment]
Ah, so you want to propagate worker pool not the set of idle workers.  That's possible with #35833 but, unless `schedule` (hence `@spawn`) is aware of such worker pool, it's incomplete.

I think it's possible to allow dynamically scoped scheduler by something like

```julia
abstract type AbstractScheduler end
struct DefaultScheduler <: AbstractScheduler end

@contextvar SCHEDULER::AbstractScheduler = DefaultScheduler()

schedule(task) = schedule(SCHEDULER[], task)

schedule(::DefaultScheduler, task) = ...  # current implementation
```

so that you can do `with_context(Base.SCHEDULER => CustomScheduler()) do ... end` to enable your scheduler.  You can even do `remotecall` in `schedule(::CustomScheduler, task)` if local pool is used up.  That's the best point to propagate local context to remote.

I'm not sure customizable scheduler can be in `Base` since it adds one more dynamic dispatch per `@spawn`.  But #35833 lets you build this as an external package.

**Comment by oschulz** [comment]
Oh, yes - sorry, I should have been more clear: My intent isn't distributed scheduling here (though we'll certainly want that too, of course), but distributed allocation/partitioning of resources like workers, data chunk (references), etc.

**Comment by c42f** [comment]
I had a nice chat to @JeffBezanson about this (summary in https://github.com/JuliaLang/julia/pull/35833#issuecomment-655872620). I think the most important part we discussed was efficiency. Whatever API we expose here, it should permit *really efficient code generation* for dynamically scoped variables in the future.

The key to good code generation is to make sure the optimizer can reason locally about the *value* of dynamically scoped variables within a function and its inlined children. I think this implies that setting of context vars should be *strictly nested* with the call stack. (Roughly speaking, that `with_context` from #35833 is the only way to to set these variables.)

With strict nesting, any child function cannot affect the value of a context var in the parent. This gives the compiler a license to do some useful optimizations:
* It can **hoist the lookup machinery and load** of a given context var to the outer scope of any function (or from an inlined function to the parent). This allows the use of context vars inside loops and inline-worthy functions without continually paying the lookup cost.
* Context vars can be **propagated without any lookup** when the hoisting of said variable hits the construct which sets that same variable (`with_context` in #35833).
* In principle, the above two optimizations can supply **precise type information** at inlined use sites, leading to many further optimizations. If we want to attempt an inferrable effects system on top of context variables I feel like this is a required optimization. (Caveat: it's not clear to me how to model this optimization in the existing optimizer in a way which lets other passes see the type information. Context vars could be modeled as implicit function arguments, but I don't know whether this can be kept under control in terms of compile time in type inference, and the interaction with inline-worthy computation worries me a bit. Having said that, I really don't know the optimizer code at all, so a lot of this is mere speculation.)

It seems to me that little of the above can be done in the presence of non-strict nesting (except, perhaps for leaf functions after inlining).

**Comment by tkf** [comment]
Thanks for the summary (here and in #35833)!

My immediate question is that "is it also reasonable to allow slow but easy-to-write code?"  Why not let the performance of the function gradually degrade when it uses non-compiler friendly constructs?  For example, I think building a concurrent-safe environment variable interface (where people can just do `ENV[key] = value`) would require something like `contextvar[] = value` interface.  You probably don't care about the performance when spawning a process.  So, using convenient syntax might be more valuable than optimizing the function.  If you care about the performance, re-structuring code to use `with_context` wouldn't be so bad (it's kind of like using `@inbounds`).  But, if you are writing (say) an HTTP service, I think it'd likely be too cumbersome as an interface.

> It seems to me that little of the above can be done in the presence of non-strict nesting (except, perhaps for leaf functions after inlining).

By non-strict nesting you mean the "in-place" interface like `contextvar[] = value` (`set!`)?  If everything in `with_context` is inlined, I suppose you can eliminate the loads and stores (maybe that's what you mean by "except, ..."')?

Also, I wonder if it's possible to accumulate the "patch" of the changes to the context variables and apply it when exiting the function (including exception)?  Since the values of the context variables are not shared across tasks, I guess this is a valid optimization (in the sense that the difference is not observable)?

**Comment by c42f** [comment]
> My immediate question is that "is it also reasonable to allow slow but easy-to-write code?" Why not let the performance of the function gradually degrade when it uses non-compiler friendly constructs?

The difficulty is that the compiler must be able to reason about code which is carefully written to be fast, and allowing slow code elsewhere might ruin some global properties which this reasoning might rely on.  Consider:

```julia
function foo(x)
    with_context(a=>x) do
        bar()
        a[]  # What is the value of `a` here? Can the compiler prove it's equal to x?
    end
end
```

If `bar()` (or any function it calls) is allowed to `set!` `a` in a way that's observable in `foo()`, this makes optimization much more difficult. It's perhaps possible in the case that `bar` is fully inlined, but this would not be true for many common use cases. Logging is a particular example — by design that calls into a potentially deep tree of non-inlinable functions.

**Comment by c42f** [comment]
> By non-strict nesting you mean the "in-place" interface like `contextvar[] = value` (`set!`)? If everything in `with_context` is inlined, I suppose you can eliminate the loads and stores (maybe that's what you mean by "except, ..."')?

Right, yes exactly this is what I meant. Essentially the nesting is about keeping side effects under control so that the compiler can apply local reasoning based just on the code within the function. Of course if everything is inlined the compiler can see all the code, but I think there's many practical cases where this won't be true.

BTW it's not just loads and stores, but any internal machinery required to look up the context var storage, eg hashing and/or traversal in the Dict/HAMT/whatever storage backend.

**Comment by tkf** [comment]
> If `bar()` (or any function it calls) is allowed to `set!` `a` in a way that's observable in `foo()`, this makes optimization much more difficult.

Oooh, I see. Thanks for the explanation!

> If we want to attempt an inferrable effects system on top of context variables I feel like this is a required optimization.

BTW, can we implement a full effect handler on top of context variables?  I wonder if fancier things like choose effect is possible?  It'd imagine what we can do with the context variables would be limited to some kind of function dispatch system.  Instead, if implementing effect system is a reasonable goal, maybe implementing the effect system _first_ and then implement the context variable as the "state effect" handler?  I have no idea if that'd be more compiler friendly, though.

(Incidentally, there is a discussion in Zulip on the performance and implementation strategies of effect handler.  It's about using delimited continuation as the building block of effect handler.  https://julialang.zulipchat.com/#narrow/stream/137791-general/topic/Performance.20and.20implementation.20strategies.20of.20algebraic.20effec)

**Comment by oschulz** [comment]
@c42f how would this play with propagating context to tasks and remote calls? My original motivation was for propagating information about available resources (workers, etc.) in a hierarchical computation. If context is propagated via stack (that should indeed be very fast), would we extract context from the stack and make it part of a closure or something like that?

**Comment by c42f** [comment]
> BTW, can we implement a full effect handler on top of context variables?

Maybe not? I've had an uneasy feeling that there's some fundamental limitations of doing it on top of context vars. But there's a gaping hole in my understanding of the CPS transform and why people use it for this stuff, so I'm not sure what the limitations are :laughing: I need to read more here (thanks for the links!) but I'd also give up some generality for efficiency...

> how would this play with propagating context to tasks and remote calls

My thought is that it conceptually "works the same everywhere". For efficiently spawning local tasks it should be effective to use a heap allocated [persistent data structure](https://en.wikipedia.org/wiki/Persistent_data_structure) (@tkf has mentioned the HAMT and I think he had a persistent variant in mind?). Then I think you can just copy a tiny fixed size handle to the current version of that data structure during task spawning. I'm not quite sure this all makes sense but it's how I'd want it to work :-)

For remote tasks we'd need to copy the context vars and serialize them, sending them to the remote side as part of invoking the remote function. With strict nesting we might get some useful optimizations here in limiting the amount of state we serialize and send?

**Comment by oschulz** [comment]
Sounds great - if that would "just work" in a transparent manner, I think we'd have something very powerful on our hands. Originally, I thought this would be only a high-level thing at the task and remote-call level - without type-stability, since fast code would always be behind some kind of dispatch, and information like "available workers" would obviously never even reach time-critical code.

But if this could become something that can reach through to deep level of code, it would obviously enable completely different kind of applications as well, as a fully-fledged resource-injection framework. Propagation of things like "default AD-engine", etc., spring to mind. And we'd have a clean way to propagate loggers & friends.

We'll have to be careful to keep people from abusing it, though (replace kwargs by context too often, etc.). :-)

**Comment by tkf** [comment]
> I'd also give up some generality for efficiency...

I agree that's a good engineering in many cases. But I suspect it'd let us implement something like ["nano-coroutine"](https://www.youtube.com/watch?v=j9tlJAqMV7U).  It'd also help implementing something like GeneratorsX.jl without a big gun like IRTools.jl.  So all these possibilities are very tempting to me :)

> @tkf has mentioned the HAMT and I think he had a persistent variant in mind?

Yes, I was thinking about the persistent/immutable data structure.

> For remote tasks we'd need to copy the context vars and serialize them, sending them to the remote side as part of invoking the remote function. With strict nesting we might get some useful optimizations here in limiting the amount of state we serialize and send?

Unfortunately, it's very easy to invoke `with_context(x => a_1TB_array, y => open(temp_file), z => pointer(...)) do ... end`.  I don't think automatic propagation to remote workers is reasonable.  I don't think strictly nested scope or persitent data structure would help with it.

In #35833, I tried as much as possible to make the API and implementation Distributed.jl-friendly so that some kind of opt-in/manual propagation is possible.  The strategy I had in mind was to get the current snapshot of the backend dictionary by `snapshot_context`, send it to remote (maybe after filtering out unsafe contents), and then start the remote execution inside `with_context(f, snapshot)`.  In a way, `snapshot_context` acts like `Base.@locals`.  I suppose this is a compiler-friendly construct and we can keep it?  So far we discussed "in-place mutation" and it seems to be a safe API to have as long as we don't mutate the context variable storage.

**Comment by oschulz** [comment]
>  I don't think automatic propagation to remote workers is reasonable. I don't think strictly nested scope or persitent data structure would help with it.

But that pretty much precludes the use case behind my original proposal: A clean way to propagate information about available resources (like workers and logging) to child-tasks and across remote calls. Do you thing we should go for using two separate mechanisms, here?

**Comment by c42f** [comment]
> But I suspect it'd let us implement something like ["nano-coroutine"](https://www.youtube.com/watch?v=j9tlJAqMV7U). It'd also help implementing something like GeneratorsX.jl without a big gun like IRTools.jl. So all these possibilities are very tempting to me :)

I watched the nano-coroutine talk a while back. I agree it's very compelling but didn't realize we might address it this way. Sounds interesting for sure.

> We'll have to be careful to keep people from abusing it, though

Yes dynamic scope is easy to abuse and can lead to confusing code. We discussed this a bit following the Go post mentioned in https://github.com/JuliaLang/julia/pull/35833#issuecomment-633754489.

> Unfortunately, it's very easy to invoke `with_context(x => a_1TB_array, y => open(temp_file), z => pointer(...)) do ... end`. I don't think automatic propagation to remote workers is reasonable.

I wondered whether we could allow it by sending all keys and poisoning keys containing "unreasonable" state at the remote end (eg, non-serializable state, or things which are huge). By "poisoning" I mean sending the key but making loading the value an error. But that does seem pretty arbitrary and kind of awful :grimacing: Maybe instead it's a per-contextvar Bool which indicates whether it's intended to be distributed? This might make sense because the library declaring the context var has a chance to know it's not going to put large or non-serializable data inside.

> I don't think strictly nested scope or persitent data structure would help with it.

Yes it doesn't help with those problems.

**Comment by tkf** [comment]
@oschulz I'm sorry if I gave you a misleading impression that I'm implementing exactly what was proposed in the OP.  I've been trying to abstract out the requirements and trying to design the system that is useful for a wide range of usecase.  We need to get the design correct and maximally extensible before actually implementing whatever possible in Distributed.jl.  Please remember there are other important usecases as listed in the OP of #35833.

Having said that, I think it'd be reasonable to have (say) `Shared` wrapper type s.t. inside `with_context(x = Shared(value)) do ... end`, `value` would be automatically propagated to the remote calls.  This should be implementable purely inside Distributed.jl and without altering anything in the core implemented in #35833.

> Maybe instead it's a per-contextvar Bool which indicates whether it's intended to be distributed?

@c42f I don't think it's possible to do this with the UUID-key based implementation like #35833 because the task-local dictionary itself doesn't know about the contextvar properties.  That's why I think moving this information to the value side (as in `Shared(value)` suggested above) would be easier to implement.

**Comment by oschulz** [comment]
> Please remember there are other important usecases as listed in the OP

Sure, of course! I was just wondering it we should split this into two different issues, or if all use cases can be covered by a common mechanism. I think that would be possible (and preferable) if we do propagate across task and remote-call boundaries.

**Comment by oschulz** [comment]
We could always check if the information is "too big" to be forwarded via remote call.

**Comment by tkf** [comment]
> I think that would be possible (and preferable) if we do propagate across task and remote-call boundaries.

It's not possible and not preferable because:

* There are Julia objects that are inherently process-local. Pointers and file descriptors are basic examples.
* It is not possible to know if a context variable holding a heap-allocated data structure is meant to capture the _identity_ or the _value_.
    * For example, if I have a `x = Vector{Vector{Int}}`, is there some code expecting `x[1]` to be mutated in-place in a specific way (i.e., this context is propagating the identity)?  Or, is the program correct if I do `x[1] = copy(x[1])` at random points (i.e., this context is propagating the value)?
* Not all objects are serialization-safe (e.g., it may need `finalizer`).
* The serialization overhead of large objects.

I have proposed various solutions to this problem. I think the discussion would be more productive if you explain why they don't work.

**Comment by oschulz** [comment]
> I have proposed various solutions to this problem. I think the discussion would be more productive if you explain why they don't work.

Uh, maybe we misunderstood each other: I'm with you for rejecting things that can't be forwarded, requiring big objects to be wrapped in a `Shared` wrapper or so, etc.

Maybe this was a misunderstanding on my side, I had the impression that you didn't want non-explicit propagation to tasks and remotes at all anymore, because you wrote "I don't think automatic propagation to remote workers is reasonable". And I was wondering how many use case would still work if context had to be propagated explicitly - since the code that distributes work to tasks and remotes (say, Transducers :-) ) will often not know about the semantics of the whole context. 

I would assume that in the future, we'll have more and more automatic/transparent multi-threaded and also multi-process code execution. So the code that "declares" the context, and the code that "consumes" the context will often not be aware of the task/remote-call barrier in between, and may not share stack. But the code that does the parallelization (say, a multi-threaded broadcast implementation) will not know/care about what's in the context - except for the parts of the context that control parallelization.

So that's why I think context must, in principle, always be automatically forwarded to spawned tasks and to remote calls. But of course we can reject/filter certain types of content, resp. require them to be wrapped appropriately - context should, in my opinion, not be abused as a data store for substantial amounts of data, and that should be discouraged.

**Comment by oschulz** [comment]
> It is not possible to know if a context variable holding a heap-allocated data structure is meant to capture the identity or the value

I guess a clean way around that is to only allow context to refer to immutable values (we do have an immutable array type somewhere, don't we?). If we do that, copies can be made as necessary, transparently, without affecting semantics.

**Comment by c42f** [comment]
> I would assume that in the future, we'll have more and more automatic/transparent multi-threaded and also multi-process code execution. So the code that "declares" the context, and the code that "consumes" the context will often not be aware of the task/remote-call barrier in between, and may not share stack. But the code that does the parallelization (say, a multi-threaded broadcast implementation) will not know/care about what's in the context - except for the parts of the context that control parallelization.

Yes this is exactly the reason that some portion of the context needs to be propagated automatically
1. The code which calls `Distributed.@spawn` exposes (distributed) parallelism
2. The library which creates the context var knows whether it can be safely distributed (if anything does!)

In general (2) is not the end user's top level application code and there's a good chance that (1) might also not be. So they definitely need to be decoupled.

I think this is why @tkf was suggesting the `Shared` wrapper. I don't like the idea of manually wrapping and unwrapping, but I suppose the context var get/set interface itself could do that.

**Comment by oschulz** [comment]
I guess in most (sane) cases, the entries of the context would be fairly small and immutable structures anyhow. I guess we can be fairly rigorous and filter everything out that can't be propagated automatically. Maybe we should actually reject everything we don't "like" during context creation/assignment already, to avoid surprises to the user later on?

**Comment by tkf** [comment]
@oschulz Thanks for the clarification.  Indeed, it looks like we have a different notion of "automatic", "explicit", etc.  To clarify, I've been using "explicit" to mean that the user does _something_ beyond the standard context variable declaration. This is maybe some kind of declarative API to tell Distributed.jl to propagate certain context variables for every remote-call (opt-in).  Or, maybe just using a low-level API to reset the remote context (manual).  If what you mean by "automatic" is what I mean by "opt-in", yes, we are actually on the same page.  Perhaps I should have mentioned "_unconditional_ propagation is not reasonable" instead of "_automatic_ propagation is not reasonable."

Concretely, I think it is reasonable to have API something like

```julia
@contextvar PROCESS_LOCAL_CONTEXT = 1   # not forwarded to remote
@shared_contextvar GLOBAL_CONTEXT = 1   # automatically forwarded to remote
```

(I think it kind of makes sense to call them `@contextobject` and `@contextvalue`.)

> I guess we can be fairly rigorous and filter everything out that can't be propagated automatically.

This is the point I'm still strongly against as I explained in the last comment https://github.com/JuliaLang/julia/issues/35757#issuecomment-656461947.  We should make API explicit and easy to understand and manipulate. Something implicit will cause trouble.

> but I suppose the context var get/set interface itself could do that.

@c42f Yeah, I think that's possible.  It can be an option to `@contextvar` or another macro like `@shared_contextvar`.  We can totally hide that there is a wrapper like `Shared`.  Or, there can be two dictionaries as backed.  No one other than Distributed.jl and Base would notice this.

**Comment by oschulz** [comment]
>  "unconditional propagation is not reasonable" instead of "automatic propagation is not reasonable."

I fully agree. I think it's perfectly fine to expect the user to declare a context variable that is to be propagated automatically in a certain way, and to restrict it to certain types of content. It's certainly good to let the user control over what should be restricted to the current process, and what should propagate beyond.

The "local" `@contextvar` still be propagated to tasks, though, right? I guess in the future, tasks will be used to much under the hood that the user will often really not even be aware of it. And it's all the same shared memory, so propagation of (almost arbitrary but immutable) content wouldn't be a problem?

**Comment by tkf** [comment]
Nice to know that we are on the same page!

> The "local" `@contextvar`

I commented this in the other issue https://github.com/JuliaLang/julia/pull/35833#issuecomment-657019038

**Comment by c42f** [comment]
> It can be an option to `@contextvar` or another macro like `@shared_contextvar`. We can totally hide that there is a wrapper like `Shared`. Or, there can be two dictionaries as backed. No one other than Distributed.jl and Base would notice this.

Nice, I think this is the way to go.

My inclination is to have options to `@contextvar` rather than having `@shared_contextvar`, as I feel like there could be other fine grained properties which define how to propagate or otherwise manage/define context variables in the future. (For example, what about GPUs? What about context vars which need code to run when new tasks spawn? Generally I feel like we should minimize the differences between in-process vs out-of-process tasks where possible — this binary distinction isn't the only way to categorize available compute.)

**Comment by tkf** [comment]
> this binary distinction isn't the only way to categorize available compute

Right, it makes sense.

> What about context vars which need code to run when new tasks spawn?

I'm glad that you are shooting for this!  It'd make it possible to implement something like parallel RNG completely in the user space.

> eg hashing and/or traversal in the Dict/HAMT/whatever storage backend.

BTW, regarding the lookup overhead, we can store `hash(uuid)` in `ContextVar` object so that it's computed only once. Traversal is still a bottleneck, though.  Anyway, it's a micro-optimization we can try at some point.

**Comment by tkf** [comment]
Speaking of parallel RNG, it'd require some thing like

```julia
@contextvar RNG_STATE = RNGState(...)
```

where `RNGState` is mutable, right?  In general, "no-`set!`" direction can simply be defeated by

```julia
@contextvar IMMUTABLE_POINTING_TO_MUTABLE = Ref(thing)
```

How does it interact with the optimization you have in mind?  Is it that it's important to make the default get-only?

**Comment by c42f** [comment]
> where RNGState is mutable [...] How does it interact with the optimization you have in mind?

I think the point here is that *context should behave like implicit arguments to child functions* from the point of view of the compiler. Then the user has a choice to make context immutable or not as necessary, and the compiler can reason about the values of these variables locally in the same way as normal function arguments.

For context like `Ref{Int}()`, the compiler still gets benefits in knowing that the address of the `Int` cannot change, even though it may need to emit a load of the `Int` after running any child functions.

**Comment by tkf** [comment]
> For context like `Ref{Int}()`, the compiler still gets benefits in knowing that the address of the `Int` cannot change, even though it may need to emit a load of the `Int` after running any child functions.

Thanks for the explanation. It makes sense.

**Comment by c42f** [comment]
> > What about context vars which need code to run when new tasks spawn?
> 
> I'm glad that you are shooting for this! It'd make it possible to implement something like parallel RNG completely in the user space.

Yes exactly! I feel like whatever we come up with here should be able to support both https://github.com/JuliaLang/julia/pull/34852 and logger context in user space with excellent efficiency. If not, we won't really have solved two of the key use cases.

Regarding mutable context, it basically has to be cloned (at the vary least) in the parent task prior to `@spawn` otherwise we'll generate data races everywhere. This is a worrying part of the design because it forces some code to run in `@spawn` even if the functions called in the new task never use that state.

**Comment by tkf** [comment]
Right, `Ref(thing)` example does not really make sense.

I think it's also important to mutate the state in the current context upon `@spawn`.  Otherwise, after

```julia
t1 = @spawn ...
t2 = @spawn ...
```

we have identical RNG state in `t1` and `t2`.

**Comment by oschulz** [comment]
> I'm glad that you are shooting for this! It'd make it possible to implement something like parallel RNG completely in the user space.

RNGs have actually been on my mind as a potentially very important use case for contexts. While we often forward RNG via an explicit parameter (and should), sometimes (e.g. in for a likelihood function that happens to need an RNG internally, but just takes the model parameters as it's input) it would be great to pass it on via context.

For parallel applications, I usually use a counter-based RNG, so that I can use a common seed and partition the parameter space in a hierarchical fashion - that does require semantic knowledge, but should be easy to do using the proposed `with_context`.

**Comment by oschulz** [comment]
> @c42f I think the point here is that context should behave like implicit arguments to child functions from the point of view of the compiler.

I think that's the ideal way to define it, semantically.

**Comment by NHDaly** [comment]
We've wanted something like this for a while too! :) Thanks for opening the issue and discussing it! 👍 

Since nothing like this exists right now, we've been toying with the (dirty) idea of (ab)using the logger to get a context that passes to child tasks, given that we currently do pass the logger through to child tasks.

For example, we were considering a thread-aware tracing framework that does something like this, even though it's clearly terrible:
```julia
using Logging

struct TraceLogger <: AbstractLogger
    span_name::String
    parent_logger::AbstractLogger
end

Logging.min_enabled_level(tl::TraceLogger) = Logging.min_enabled_level(tl.parent_logger)
Logging.shouldlog(tl::TraceLogger,args...) = Logging.shouldlog(tl.parent_logger, args...)
Logging.handle_message(tl::TraceLogger,args...) = Logging.handle_message(tl.parent_logger, args...)

ancestor_trace(::Any) = ""
ancestor_trace(tl::TraceLogger) = "$(tl.span_name), $(ancestor_trace(tl.parent_logger))"


function with_span(f, name)
    with_logger(TraceLogger(name, Logging.current_logger())) do
        info = @timed f()
        @info "Finished $(ancestor_trace(Logging.current_logger())): $(info.time)"
        return info.value
    end
end
```
```julia
julia> with_span("a") do 
           @sync begin
               @async begin
                   with_span("b") do
                       @info "HI"
                   end
               end
               @async begin
                   with_span("C") do
                       @info "BYE"
                   end
               end
           end
       end
[ Info: HI
[ Info: BYE
[ Info: Finished b, a, : 0.052175791
[ Info: Finished C, a, : 0.000226425
[ Info: Finished a, : 0.101278193
Task (done) @0x00000001136d58d0
```
:)

Anyway, yeah, sign us up as another interested party!

**Comment by oschulz** [comment]
> EDIT: Oh, except i've just realized that this thread is about communication across distributed tasks, not necessarily about communication across multithreaded tasks? Or is it covering both?

**Comment by oschulz** [comment]
> EDIT: Oh, except i've just realized that this thread is about communication across distributed tasks, not necessarily about communication across multithreaded tasks? Or is it covering both?

Definitely both, also across local tasks. And since my original proposal, @tfk, @c42f and @JeffBezanson have taken this idea even further than I had envisioned originally - something that could possibly also be used in time-critical code. I think this could potentially become an extremely powerful mechanism.

**Comment by vchuravy** [comment]
In #50958 I made the semi-intentional decision not to address the remote-call part of this proposal.
But a distributed framework could choose to propagate a subset of scoped variables across a remote-call interface.

I have started a small prototype for snapshotting in https://github.com/vchuravy/ScopedValues.jl/pull/6, but I won't be including this into the Base proposal for now. The implementation thereof can live in a package. 

(we might also be able to implement a "RemoteScope" on-top of local scopes, but I haven't thought to hard about this).

**Comment by vchuravy** [comment]
The issue is that copying a scope is a very heavy operation. It's something I decidedly wouldn't want to do on every rpc.

Now a framework should be able to define that it wants to propagate scope with its rpc and snapshot only relevant pieces.
Right now for me the design uncertainty with Distributed.jl is large. It may be that we want something like a RemoteSopedValue akin to a RemoteRef,
but the cost must be local and not global across the program.

I certainly wouldn't want to send CuContext or CuDevice automatically across the wire.

**Comment by oschulz** [comment]
> The issue is that copying a scope is a very heavy operation.

Hm, yes, there is that ... hard to control how much people will put in there.

**Comment by vtjnash** [comment]
Seems covered by https://github.com/JuliaLang/julia/pull/50958 for the main case above of `@async` and `Threads.@spawn`
